### PR TITLE
Initial restructure to support multiple target architectures

### PR DESCRIPTION
**Pull request by mikee47**
As per discussion #1333.

Rearrange files and fix #includes:

* Rename `SmingCore` -> `Core`
* Rename `SmingCore/Platform` -> `Platform`
* Rename `system` -> `System`
* Move `Services/libb64` -> `Components`
* Move `Services/WebHelpers` -> `Core/Network`
* Remove `apptest`
* Move Esp8266-specific code:
	* Move most `Platform` sources -> `Arch/Esp8266/Platform`
	* Move some `SmingCore` sources -> `Arch/Esp2866/Core`
	* Rename `compiler` -> `Arch/Esp8266/Compiler`
	* Rename `tools` -> `Arch/Esp8266/Tools`
	* Move `SmingCore/Network/rBootHttpUpdate` -> `Arch/Esp2866/Core/Network`
	* Move `custom_heap` -> `Arch/Esp8266/Components`
	* Move some of `system` files -> `Arch/Esp8266/Components/esp8266` (reflects IDF layout)
	* Rename `appinit/user_main.cpp` -> `Arch/Esp8266/Components/esp8266/startup.cpp`
	* Rename `gdb` -> `Arch/Esp8266/Components/gdbstub`
	* Rename `appspecific/gdb` -> `Arch/Esp8266/Components/gdbstub/appcode`
	* Rename `Services/FATFS` -> `Arch/Esp8266/Components/fatfs`
	* Move `system/include/rboot-integration.h`, `appspecific/rboot/overrides.c` -> `Arch/Esp8266/Components/rboot/appcode`
	* Rename `Services/SpifFS` -> `Arch/Esp8266/Components/spiffs`
	* Move `include/user_config.h` -> `Arch/Esp8266/System/include`

Makefile revisions

* Remove `Makefile-project.mk`
* Rename `Makefile-rboot.mk` -> `Makefile-app.mk`
* Retain `Makefile-rboot.mk` (redirects to `Makefile-app.mk`) for backward compatibility
* Separate common script used by both Sming and App
	* `build.mk` contains host environment setup
	* Add directory definitions `ARCH_BASE`, `ARCH_SYS`, `ARCH_CORE`, `ARCH_COMPONENTS` and `ARCH_TOOLS`
	* `modules.mk` contains rule creation script
* Separate-out ESP8266-specific script:
	* `Arch/*/build.mk` included by `build.mk` file
	* `Arch/*/sming.mk` included by `Makefile`
	* `Arch/*/app.mk` included by `Makefile-app.mk`
* Change build output directory to `Arch/*/out`
* Pull related variable definitions and rules together
* Define macros to replace repetitive script: endeavour to name files or directories once only
* Ignore errors in all clean operations
* Use `TOOLS`, `TOOLS_CLEAN` and `CLEAN` variables containing pre-requisites for `tools`, `tools-clean` and `clean` targets
* Simplify user library building by adding to `LIBS`, each library gets an additional shorthand make target e.g. `make axtls`.
* Distribute PHONY declarations
* Add `decode-stacktrack` rule to application makefile
* Add `help` target to extract formatted makefile comments
* Add `list-config` target

Submodule handling

* third-party libraries: `rBoot`, `new-pwm`, `axtls-8266`, `umm_malloc`, `esp-open-lwip`, `lwip2` -> `Arch/Esp8266/Components`
* Move `.patch` file into related submodule parent directory
* Internal SDK moved into `Components/Sdk` and updated to use unpatched Version 3.0 release
* Move remaining `third-party` submodules into `Components`
* Revise Makefiles so that only those submodules required for a build are updated/patched and built
	* Add `SUBMODULES` variable to specify which submodules are required
	* Use `.submodule` file as consistent pre-requisite for update/patching of all submodules
	* Make `ARDUINO_LIBRARIES` a public variable to optionally restrict which libraries get loaded and built
	* Add `submodules` and `submodules-clean` rules to replace `third-party` and `libraries` rules

Other

* Add `building.md`
* Add help documentation to makefiles, with new `help` and `list-config` targets
* Update coding style directories and Travis checks
* Update Doxygen directories

Fixes:

* Add `libpwm_open` to `CUSTOM_TARGETS` so it builds with framework

Bugs found:

* Testing with `ENABLE_CUSTOM_LWIP=0` fails because of static definitions in `mem_manager.h` - also happens with existing build

### CODE REVIEW

**Comment by slaff** [comment]
> * `Platform` -> `Platform/Esp8266/Platform`

Wouldn't it be better to rename it to `Platform/Esp8266`? Without one more `Platform` at the end.

> * `custom_heap` -> `Platform/Esp8266/System/components/custom_heap`

Why not `Platform/Esp8266/Components/custom_heap` ? Without the `System` part.  The `Components` directory for every platform should be the place to store third-party libraries, components, etc. For example `Platform/Esp8266/System/components/spiffs` should be `Platform/Esp8266/Components/spiffs` 

> * `appinit/user_main.cpp` -> `Platform/Esp8266/System/components/esp8266/startup.cpp`

Why not `Platform/Esp8266/AppInit/startup.cpp` ? This is shorter and should be one of the files that every platform must implement.

> * Move Esp8266-specific files from `system` into `Platform/Esp8266/System/components/esp8266`

How about `Platform/Esp8266/System/` ? 

> * `gdb` -> `Platform/Esp8266/System/components/gdbstub`

Also `Platform/Esp8266/Components/gdbstub`.

> Bugs found:
> 
> * Testing with `ENABLE_CUSTOM_LWIP=0` fails because of static definitions in `mem_manager.h` - also happens with existing build

Interesting. I have double checked this and you are right. That's strange because it used to work... but it is not a big issue because I guess no one has used the Espressif's LWIP since years.

**Comment by mikee47** [comment]
I've noticed that this PR consistently uses about 60 bytes more IRAM. I tracked this down to the `Cache_Read_Enable_New()` function in `rboot-big-flash.c`, which produces different code with `-mforce32` enabled.
The previous makefiles only enabled the `-mforce32` option for building the framework - it was conspicuously absent from project builds - but now both framework and app build with the same settings (defined in `build.mk`) which means -mforce32 is applied consistently.

**Comment by slaff** [comment]
... also `Sming/apptest` folder and its content can go away. Every platform can have its owns `Tests` folder where different unit tests can be stored. But tests and test frameworks is a whole new big topic that I don't want to mix in this PR.

`Sming/third-party` can be renamed to `Sming/Components` to be consistent with the platform folder structure and will be used for third-party code that is platform agnostic. `Sming/Components` and `Sming/Platform/X/Components` should use the same patching mechanism and organization.

**Comment by mikee47** [comment]
>
>
> > * `Platform` -> `Platform/Esp8266/Platform`
> 
> Wouldn't it be better to rename it to `Platform/Esp8266`? Without one more `Platform` at the end.

We've kind of appropriated `Platform` now to mean something a bit different - perhaps the code previously living in `Platform` may sit more comfortably in `System` instead?

It's also going to be more important to 'namespace' header files, thus:

  "AccessPoint.h"               ->    <Platform/AccessPoint.h> or <System/AccessPoint.h>
  "WiringFrameWorkIncludes.h"   ->    <Wiring/Framework.h>
  "gdb_hooks.h"                 ->    <gdb/hooks.h>
  <gdb_syscall.h>               ->    <gdb/syscall.h>
  "gdb/gdbstub-internal.h"      ->    <gdbstub/internal.h>
  "Services/SpifFS"             ->    <spiffs/spiffs.h>

This means `Platform/Esp8266` cannot contain any code in the root directory.

>
> 
> > * `custom_heap` -> `Platform/Esp8266/System/components/custom_heap`
>
> Why not `Platform/Esp8266/Components/custom_heap` ? Without the `System` part. The `Components` directory for every platform should be the place to store third-party libraries, components, etc. For example `Platform/Esp8266/System/components/spiffs` should be `Platform/Esp8266/Components/spiffs`

I chose to use lower-case `components` for consistency with the ESP-IDF - As there isn't a non-OS IDF framework for the ESP8266 this is the substitute. For the Esp32 `components` lives in the SDK.
Using `Components` as well may cause confusion or other issues, especially with Windows as it's not case sensitive. Regarding this and your point below about `third-party`, how about something different like `ThirdParty`, `Shared`, `External`, `Import`, etc?
(Notwithstanding, if the intention is to supplement IDF modules then it would go into `components`.)

>
> 
> > * `appinit/user_main.cpp` -> `Platform/Esp8266/System/components/esp8266/startup.cpp`
> 
> Why not `Platform/Esp8266/AppInit/startup.cpp` ? This is shorter and should be one of the files that every platform must implement.

Our `startup.cpp` contains lots of low-level stuff which the IDF takes care of, which is why I chose `components/esp8266' to be an appropriate location.
The ESP8266 RTOS IDF implements this in `components/esp8266/source/startup.c`, and the ESP32 in `components/esp32/cpu_start.c`. We'd need to implement `app_main()` and call `init()`, that could could probably just go into `Platform/Esp32/System/app_main.cpp`.

>
> 
> > * Move Esp8266-specific files from `system` into `Platform/Esp8266/System/components/esp8266`
> 
> How about `Platform/Esp8266/System/` ?

IDF

>
>
> > * `gdb` -> `Platform/Esp8266/System/components/gdbstub`
> 
> Also `Platform/Esp8266/Components/gdbstub`.

IDF

>
> 
> ... also `Sming/apptest` folder and its content can go away. Every platform can have its owns `Tests` folder where different unit tests can be stored. But tests and test frameworks is a whole new big topic that I don't want to mix in this PR.

OK.

>
> 
> `Sming/third-party` can be renamed to `Sming/Components` to be consistent with the platform folder structure and will be used for third-party code that is platform agnostic. `Sming/Components` and `Sming/Platform/X/Components` should use the same patching mechanism and organization.

The only thing in `third-party` which isn't platform-agnostic is `ESP8266_NONOS_SDK`. Feels a bit weird having that as a `component`.. I notice it's also (slightly) patched - is there any reason we're not using the current 3.0.0 release? I think it would be preferable to use an un-patched SDK and amend makefiles if necessary to accommodate the change from `esp_init_data_default_v08.bin`. We might also consider putting the SDK itself outside of the Sming repo...

>
> 
> > Bugs found:
> > 
> > * Testing with `ENABLE_CUSTOM_LWIP=0` fails because of static definitions in `mem_manager.h` - also happens with existing build
> 
> Interesting. I have double checked this and you are right. That's strange because it used to work... but it is not a big issue because I guess no one has used the Espressif's LWIP since years.

Both `esp-open-lwip` and `lwip2` have `#include "mem_manager.h` commented-out. If we do the same it builds, but haven't done any testing on a real device.

**Comment by mikee47** [comment]
How's about this:

`third-party/http_parser`  -> `SmingCore/Network/Http/http_parser`
`third-party/mqtt-codec`  -> `SmingCore/Network/Mqtt/mqtt-codec`
`third-party/libyuarel`  -> `SmingCore/Network/Http/libyuarel`
`third-party/spiffs`  -> `Platform/Esp8266/System/components/spiffs/spiffs`
`third-party/ws_parser`  -> `SmingCore/Network/Http/Websocket/ws_parser`

That just leaves `ESP8266_NONOS_SDK` to deal with.

There's also `Services` which feels a bit out of place.

`Services/libb64` -> `SmingCore/Network/libb64`
`Services/WebHelpers` -> `SmingCore/Network/WebHelpers` 

Not sure about these:

`Services/HexDump` ?
`Services/CommandProcessing` ?
`Services/Profiling` -> `System` ?
`Services/Yeelight` ?

**Comment by mikee47** [comment]
Also, `Libraries/ArduinoJson` appears to be the only library referred to within the framework (by `JsonObjectStream`). That would suggest `SmingCore/Data` would be a better place for it?

**Comment by slaff** [comment]
> third-party/http_parser -> SmingCore/Network/Http/http_parser
third-party/mqtt-codec -> SmingCore/Network/Mqtt/mqtt-codec
third-party/libyuarel -> SmingCore/Network/Http/libyuarel

Better not. We should not mix our own-code with third-party code. Because it has its own coding styles, own compiler directives, etc. It would be much better to store external, third-party, libraries or components in strict predefined folder(s), for example 
- "Sming/Components" for all platform-agnostic third-party libraries, as `mqtt-codec` for example
- "Sming/Platform/X/Components" for all platform specific third-party libraries. 

> That just leaves ESP8266_NONOS_SDK to deal with.

ESP8266_NONOS_SDK goes to `Sming/Platform/Esp8266/Components/Sdk/

> Services/libb64 -> SmingCore/Network/libb64

No, that goes to `Sming/Components/libb64`

> Services/WebHelpers -> SmingCore/Network/WebHelpers

Yes, that sounds like a good location.

> Not sure about these:
> Services/HexDump ?

Maybe `SmingCore/Data/HexDump`

> Services/CommandProcessing ?

Maybe `SmingCore/CommandProcessing`

> Services/Profiling -> System ?

Hm... You've got me...

> Services/Yeelight ?

SmingCore/Drivers/Yeelight ?

**Comment by slaff** [comment]
> Also, Libraries/ArduinoJson appears to be the only library referred to within the framework (by JsonObjectStream). That would suggest SmingCore/Data would be a better place for it?

Leave Libraries/ArduinoJson in its place. It is part of Sming but also an Arudino Library. No need to move it to Core or Data. At the moment at least.

> We might also consider updating from V5.1.1 to version 6?

I will take a look and see if it is a lot of effort to migrate. Hopefully not.

**Comment by mikee47** [comment]
As a way of dealing with the `Platform/Esp8266/Platform` issue, we could instead use `Arch`, so it would be `Arch/Esp8266/Platform`. We'd use `SMING_ARCH` to define the architecture, but `Platform` would retain its original meaning.

**Comment by slaff** [comment]
I like both `Arch` and `Target` but would go for `Arch` just because it is shorter. So let it be `Arch` :)

**Comment by mikee47** [comment]
At some stage we might introduce sub-architectures, something akin to 'boards' in Arduino, but perhaps more flexible so various system-dependent stuff can be more easily configured. All that is probably much easier to deal with in CMake though...

**Comment by mikee47** [comment]
With the above changes, we've got quite a clean set of 'root' directories:

```
Sming/Arch
Sming/Components
Sming/Core
Sming/Libraries
Sming/Platform
Sming/Services
Sming/Wiring
```

Note that `Platform` has been moved out of `Core` - it makes some sense to do this as it largely contains platform-specific code - but if you'd prefer it moved back to `Core/Platform` let me know.

I think it probably best to leave `CommandProcessing`, `HexDump`, `Profiling` and `Yeelight` where they are (in `Services`) for now - it will probably become clearer at some stage in development where they should actually go (if indeed they need moving at all).

**Comment by slaff** [comment]
> Note that Platform has been moved out of Core - it makes some sense to do this as it largely contains platform-specific code - but if you'd prefer it moved back to Core/Platform let me know.
...
I also have further changes to the makefiles, the goal being to document and clarify the existing build system to make it easier to migrate (e.g. to CMake), rather than adding anything new.

Perfect. 

> I think it probably best to leave CommandProcessing, HexDump, Profiling and Yeelight where they are (in Services) for now - it will probably become clearer at some stage in development where they should actually go (if indeed they need moving at all).

Yes, I agree with you. 

>> We might also consider updating from V5.1.1 to version 6?
> I will take a look and see if it is a lot of effort to migrate. Hopefully not.

Meanwhile the migration to ArduinoJson v6 is ready for testing  #/1677. The API looks more elegant and intuitive. But we should check also the memory usage.

File: Sming/Makefile-app.mk [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+.PHONY: terminal
+terminal: kill_term ##Open the serial terminal
+	$(TERMINAL)
+
+.PHONY: gdb
+gdb: kill_term ##Run the debugger console
+	$(GDB)
+
+.PHONY: decode-stacktrace
+decode-stacktrace: ##If a crash occurred, use this option then cut & paste dump text as prompted
```
Comment (slaff) [inline Sming/Makefile-app.mk]: > $(Q) python $(ARCH_TOOLS)/decode-stacktrace.py $(TARGET_OUT_0)

decode-stacktrace.py accepts optional second parameter that is the file containing the stack trace. Would it be possible to add optional parameter here too. Something like:

```
decode-stacktrace:
    ...
    $(Q) python $(ARCH_TOOLS)/decode-stacktrace.py $(TARGET_OUT_0) $(TRACE)
```

and usage

```
make decode-stacktrace TRACE=/path/to/crash.stack
```

File: Sming/Arch/Esp8266/Tools/decode-stacktrace.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def extractAddresses(data):
         pipe.communicate("\n".join(extractAddresses(data)).encode('ascii'))
     else:
         while True:
-            data = input()
+            data = raw_input()
```
[the remark this answers is not in the document]
Comment (mikee47) [reply Sming/Arch/Esp8266/Tools/decode-stacktrace.py]: It broke for me on Windows with `input()` (python 2.7), but I'll take another look and do more tests on both platforms.

File: Sming/build.mk [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+				group = $$2; \
+				groups[group] = group; \
+			} /^[a-zA-Z_-]+:.*?##/ { \
+				targets[$$1, group] = $$2; \
+			} \
+			END { \
+				groupCount = asort(groups); \
+				for (g = 1; g <= groupCount; ++g) { \
+					printf "\n\033[1m%s\033[0m\n", groups[g]; \
+					targetCount = asorti(targets, idx); \
```
[the remark this answers is not in the document]
Comment (mikee47) [reply Sming/build.mk]: Could you try using `gawk` instead and see if that fixes it?

File: samples/SmtpClient/app/application.cpp [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #include <user_config.h>
-#include "SmingCore/SmingCore.h"
-#include "SmingCore/Network/SmtpClient.h"
+#include <SmingCore.h>
+#include <Network/SmtpClient.h>
```
[the remark this answers is not in the document]
Comment (mikee47) [reply samples/SmtpClient/app/application.cpp]: Once the new file/folder layout has been fixed, before committing I'll need to do the following:

* Fix `#include` guard naming
* Update all `#include`s to use the recommended/namespaced convention

[the remark this answers is not in the document]
Comment (slaff) [reply samples/SmtpClient/app/application.cpp]: > Fix #include guard naming

Or just use [#pragma once](https://en.wikipedia.org/wiki/Pragma_once). And then you will not need to fix them in the future.

**Comment by slaff** [comment]
> With the above changes, we've got quite a clean set of 'root' directories:
> 
> ```
> Sming/Arch
> Sming/Components
> Sming/Core
> Sming/Libraries
> Sming/Platform
> Sming/Services
> Sming/Wiring
> ```
> 

I guess I forgot to mention that I like a lot the new structure. IMHO it is more intuitive that it used to be. So definitely a good way forward.

**Comment by mikee47** [comment]
Something else we need to deal with is `user_config.h` which is heavily architecture-depdendent.
It's included for (almost) all of the sample applications (`LiveDebug` doesn't have it). They are all identical, except `Basic_NFC`, `Accelerometer_MMA7455` and `LED_YeelightBulb` which override the baud rate to 115200 - I suspect that's an error.
I propose we remove all of these files, thus picking up the defaults in `Arch/Esp8266/System/include/user_config.h`. Other architectures may or may not require such a file.
Existing user applications should remove their `user_config.h`, although we need to consider what to do if users may have made any changes to the file.

I guess the characteristic of the file is that it (usually) gets #included ahead of anything else so is a good place to put overrides. How about this as a template for users to follow:

```
#pragma once
#include_next <user_config.h>

<< User Customisations here >>
```

**Comment by slaff** [comment]
Seems good to me. But I would prefer to have it as a separate PR so that it is, as you wrote once, "easier to digest".

File: Sming/build.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (slaff) [reply Sming/build.mk]: Hm.. I somehow have mawk installed. We should inform the developers/users to use gawk similar to that post https://askubuntu.com/questions/406480/download-awk-to-ubuntu

[the remark this answers is not in the document]
Comment (mikee47) [reply Sming/build.mk]: I don't want to add additional dependencies un-necessarily, especially as we'll be doing all this via `CMake` and `KConfig` at some point (I hope!). Maybe if I just dump the output from `awk` into a make variable before sorting, etc....

**Comment by mikee47** [comment]
The `ARDUINO_LIBRARIES` variable I kept from the previous makefile, but perhaps `SMING_LIBRARIES` would be better?

**Comment by slaff** [comment]
> The ARDUINO_LIBRARIES variable I kept from the previous makefile, but perhaps SMING_LIBRARIES would be better?

Leave it as it is for the moment.

File: Sming/Arch/Esp8266/Tools/decode-stacktrace.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikee47) [reply Sming/Arch/Esp8266/Tools/decode-stacktrace.py]: `sys.stdin.readline()` should work for all python versions

[the remark this answers is not in the document]
Comment (slaff) [reply Sming/Arch/Esp8266/Tools/decode-stacktrace.py]: > sys.stdin.readline() should work for all python versions

Then let's go for it.

File: Sming/build.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikee47) [reply Sming/build.mk]: Not getting anywhere with this, are you OK to leave as-is?

[the remark this answers is not in the document]
Comment (mikee47) [reply Sming/build.mk]: I'll ditch the sorting, grouping is sufficient I think.

File: Sming/building.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
+# Sming makefiles
```
Comment (slaff) [inline Sming/building.md]: There are newly added files that have Windows line endings. Please, use `dos2unix` or some other tool to convert them. I saw that issue in this file - building.md, other md files and some of the new mk files.
image

File: Sming/SmingCore/SmingCore.h [hunk 6]
```diff
@@ -1,67 +1,3 @@
...
-#include "Data/Stream/TemplateFileStream.h"
-
-#include "DateTime.h"
-
-#include "../Services/FATFS/ff.h"
-#include "../Services/Yeelight/YeelightBulb.h"
-
-#endif /* _SMING_CORE_H_ */
+#include <Core/SmingCore.h>
+#pragma warning "Please update to #include <SmingCore.h>"
```
Comment (slaff) [inline Sming/SmingCore/SmingCore.h]: Put the text on one line. Should be more readable.

````
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/SmingCore/SmingCore.h:2:0: warning: ignoring #pragma warning  [-Wunknown-pragmas]
 #pragma warning "Please update to #include <SmingCore.h>"
 ^
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/SmingCore/SmingCore.h:3:0: warning: ignoring #pragma warning  [-Wunknown-pragmas]
 #pragma warning "See https://github.com/SmingHub/Sming/wiki/migrate-v3.8-v4.0"
```

File: samples/DFPlayerMini/app/application.cpp [hunk 7]
```diff
@@ -1,5 +1,5 @@
 #include <user_config.h>
-#include <SmingCore/SmingCore.h>
+#include <SmingCore.h>
```
[the remark this answers is not in the document]
Comment (mikee47) [reply samples/DFPlayerMini/app/application.cpp]: I note you renamed the Wiki page for this, now the link is broken!

[the remark this answers is not in the document]
Comment (slaff) [reply samples/DFPlayerMini/app/application.cpp]: I will take care of this.

**Pull request merged**